src/main/scala/coreplex/Ports.scala

```diff
@@ -0,0 +1,289 @@
+case class SlavePortParams(beatBytes: Int, idBits: Int, sourceBits: Int)
+case object ExtIn extends Field[SlavePortParams]
+
+///// The following traits add ports to the sytem, in some cases converting to different interconnect standards
+
+/** Adds a port to the system intended to master an AXI4 DRAM controller. */
+trait HasMasterAXI4MemPort extends HasMemoryBus {
+  val module: HasMasterAXI4MemPortModuleImp
+
+  private val params = p(ExtMem)
+  private val device = new MemoryDevice
+
+  val mem_axi4 = AXI4BlindOutputNode(Seq.tabulate(nMemoryChannels) { channel =>
+    val base = AddressSet(params.base, params.size-1)
+    val filter = AddressSet(channel * cacheBlockBytes, ~((nMemoryChannels-1) * cacheBlockBytes))
+
+    AXI4SlavePortParameters(
+      slaves = Seq(AXI4SlaveParameters(
+        address       = base.intersect(filter).toList,
+        resources     = device.reg,
+        regionType    = RegionType.UNCACHED,   // cacheable
+        executable    = true,
+        supportsWrite = TransferSizes(1, cacheBlockBytes),
+        supportsRead  = TransferSizes(1, cacheBlockBytes),
+        interleavedId = Some(0))),             // slave does not interleave read responses
+      beatBytes = params.beatBytes)
+  })
+
+  val converter = LazyModule(new TLToAXI4(params.beatBytes))
+  val trim = LazyModule(new AXI4IdIndexer(params.idBits))
+  val yank = LazyModule(new AXI4UserYanker)
+  val buffer = LazyModule(new AXI4Buffer)
+
+  memBuses.map(_.outwardNode).foreach { case mbus =>
+    converter.node := mbus
+    trim.node := converter.node
+    yank.node := trim.node
+    buffer.node := yank.node
+    mem_axi4 := buffer.node
+  }
+}
+
+/** Common io name and methods for propagating or tying off the port bundle */
+trait HasMasterAXI4MemPortBundle {
+  implicit val p: Parameters
+  val mem_axi4: HeterogeneousBag[AXI4Bundle]
+  val nMemoryChannels: Int
+  def connectSimAXIMem(dummy: Int = 1) = {
+    if (nMemoryChannels > 0) Module(LazyModule(new SimAXIMem(nMemoryChannels)).module).io.axi4 <> mem_axi4
+  }
+}
+
+/** Actually generates the corresponding IO in the concrete Module */
+trait HasMasterAXI4MemPortModuleImp extends LazyMultiIOModuleImp with HasMasterAXI4MemPortBundle {
+  val outer: HasMasterAXI4MemPort
+  val mem_axi4 = IO(outer.mem_axi4.bundleOut)
+  val nMemoryChannels = outer.nMemoryChannels
+}
+
+/** Adds a AXI4 port to the system intended to master an MMIO device bus */
+trait HasMasterAXI4MMIOPort extends HasSystemBus {
+  private val config = p(ExtBus)
+  private val device = new SimpleBus("mmio", Nil)
+  val mmio_axi4 = AXI4BlindOutputNode(Seq(AXI4SlavePortParameters(
+    slaves = Seq(AXI4SlaveParameters(
+      address       = List(AddressSet(BigInt(config.base), config.size-1)),
+      resources     = device.ranges,
+      executable    = true,                  // Can we run programs on this memory?
+      supportsWrite = TransferSizes(1, 256), // The slave supports 1-256 byte transfers
```

Comment:
It's theoretically a property of the external bus you're trying to connect to. You could drive traffic to it from the FrontPort, so it can be arbitrarily large. I'll change it to be a parameter.